src/main/java/reactor/core/publisher/MonoFilterWhen.java

```diff
@@ -45,7 +51,269 @@
 
 	@Override
 	public void subscribe(Subscriber<? super T> s) {
-		source.subscribe(new FluxFilterWhen.FilterWhenSubscriber<>(s,
-				asyncPredicate, QueueSupplier.XS_BUFFER_SIZE));
+		source.subscribe(new MonoFilterWhenSubscriber<>(s, asyncPredicate));
+	}
+
+	static final class MonoFilterWhenSubscriber<T> extends Operators.MonoSubscriber<T, T> {
+
+		final Function<? super T, ? extends Publisher<Boolean>> asyncPredicate;
+
+		Subscription upstream;
+
+		volatile FilterWhenInner asyncFilter;
+		volatile Throwable       error;
+		volatile int             filterState;
+
+		static final AtomicReferenceFieldUpdater<MonoFilterWhenSubscriber, FilterWhenInner> ASYNC_FILTER =
+				AtomicReferenceFieldUpdater.newUpdater(MonoFilterWhenSubscriber.class, FilterWhenInner.class, "asyncFilter");
+		
+		static final AtomicReferenceFieldUpdater<MonoFilterWhenSubscriber, Throwable> ERROR =
+				AtomicReferenceFieldUpdater.newUpdater(MonoFilterWhenSubscriber.class, Throwable.class, "error");
+
+		static final FilterWhenInner INNER_CANCELLED = new FilterWhenInner(null, false);
+
+		static final int STATE_FRESH         = 0; //the source value is fresh
+		static final int STATE_RUNNING       = 1; //the filter is running asynchronously
+		static final int STATE_RESULT        = 2; //the output Mono is resolved
+
+		MonoFilterWhenSubscriber(Subscriber<? super T> actual, Function<? super T, ? extends Publisher<Boolean>> asyncPredicate) {
+			super(actual);
+			this.asyncPredicate = asyncPredicate;
+		}
+
+		@Override
+		public void onSubscribe(Subscription s) {
+			if (Operators.validate(upstream, s)) {
+				upstream = s;
+				actual.onSubscribe(this);
+				s.request(1);
+			}
+		}
+
+		@Override
+		public void onNext(T t) {
+			if (filterState == STATE_FRESH) {
+				setValue(t);
+				Publisher<Boolean> p;
+
+				try {
+					p = Objects.requireNonNull(asyncPredicate.apply(t),
+							"The asyncPredicate returned a null value");
+				}
+				catch (Throwable ex) {
+					Exceptions.throwIfFatal(ex);
+						//TODO not really needed as we don't expect to get there if the Mono source errors
```

Comment:
I think we should be able to ditch the ERROR field and forward errors directly, as there shouldn't be a case where an error can happen both in the source / predicate application AND the inner subscriber... however the `filterState` is still needed, otherwise upstream will call `onComplete` and that method will not detect an error was propagated.